src/main/java/de/tum/p2p/api/netty/forwarding/OnionTunnelBuildMessage.java

```diff
@@ -0,0 +1,72 @@
+package de.tum.p2p.api.netty.forwarding;
+
+import de.tum.p2p.Peer;
+import de.tum.p2p.proto.message.MessageType;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.nio.ByteBuffer;
+import java.security.KeyFactory;
+import java.security.PublicKey;
+import java.security.spec.X509EncodedKeySpec;
+
+import static de.tum.p2p.util.ByteBuffers.randPadRemaining;
+
+@ToString
+@EqualsAndHashCode(callSuper = true)
+public class OnionTunnelBuildMessage extends OnionApiMessage {
+
+    private Peer destinationPeer = null;
+
+    public OnionTunnelBuildMessage(Peer destinationPeer) {
+        super(MessageType.ONION_TUNNEL_BUILD);
+        this.destinationPeer = destinationPeer;
+    }
+
+    public static OnionTunnelBuildMessage me(Peer destinationPeer) { return new OnionTunnelBuildMessage(destinationPeer); }
+
+    public Peer getDestinationPeer() {
+        return destinationPeer;
+    }
+
+    @Override
+    protected ByteBuffer writeMessage(ByteBuffer typedMessageBuffer) {
+        return randPadRemaining(typedMessageBuffer); // Onion Cover message carries no payload
+    }
+
+    public static OnionTunnelBuildMessage fromBytes(ByteBuffer buffer) throws Exception {
+        int minSize = 28;
+        if (buffer.remaining() < minSize) {
+            throw new Exception("Exceeds size of message");
+        } else {
+            buffer.getShort();
+            short port = buffer.getShort();
+
+            PublicKey targetHostkey;
+            byte[] addressBytes;
+            try {
+                addressBytes = new byte[4];
+                targetHostkey = KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(buffer.get(addressBytes).array()));
```

Comment:
this looks completely broken. according to spec, first goes RESERVED[15] (NOT SHORT[16]), IP ver flag[1], port[16], ip[32/64] and only then key, whereas you are parsing port first.
1) `buffer.position(MessageType.BYTES + Short.BYTES - 1)` // skip type, and RESERVED[15]  (but it's better to check the type `if (MessageType.fromBytes(buffer.getShort()) != MessageType.ONION_TUNNEL_BUILD) throw ...`
2) read ip ver flag `val ipver = buffer.get()`
3) parse port 
4) parse ip address according to `ipver` parsed
5) parse key

PS: for parsing keys use Keys.parsePublicKey(bytes)
PS2: see de.tum.p2p.proto.message.onion.forwarding.composite.TunnelConnect#fromBytes for ip and key parsing example